File: payment_backend/firebase_calls.py
import firebase_admin
from firebase_admin import credentials, firestore
from datetime import datetime
from typing import Union, Literal, Dict, Any

# Initialize Firebase
cred = credentials.Certificate('./assets/service-account.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

SubscriptionStatus = Literal['PENDING', 'ACTIVE', 'EXPIRED']
NonSubscriptionStatus = Literal['PENDING', 'COMPLETE', 'CANCELED']
IAPSource = Literal['google_play', 'app_store']

class BasePurchase:
    def __init__(self, iapSource: IAPSource, orderId: str, productId: str, userId: str, purchaseDate: datetime):
        self.iapSource = iapSource
        self.orderId = orderId
        self.productId = productId
        self.userId = userId
        self.purchaseDate = purchaseDate

class SubscriptionPurchase(BasePurchase):
    def __init__(self, iapSource: IAPSource, orderId: str, productId: str, userId: str, purchaseDate: datetime, expiryDate: datetime, status: SubscriptionStatus):
        super().__init__(iapSource, orderId, productId, userId, purchaseDate)
        self.type = 'SUBSCRIPTION'
        self.expiryDate = expiryDate
        self.status = status

class NonSubscriptionPurchase(BasePurchase):
    def __init__(self, iapSource: IAPSource, orderId: str, productId: str, userId: str, purchaseDate: datetime, status: NonSubscriptionStatus):
        super().__init__(iapSource, orderId, productId, userId, purchaseDate)
        self.type = 'NON_SUBSCRIPTION'
        self.status = status

Purchase = Union[SubscriptionPurchase, NonSubscriptionPurchase]

class FirebaseCalls:
    def __init__(self, firestore_client):
        self.firestore = firestore_client

    async def create_or_update_purchase(self, purchase_data: Purchase):
        purchases = self.firestore.collection('purchases')
        purchase_id = f"{purchase_data.iapSource}_{purchase_data.orderId}"
        purchase_ref = purchases.document(purchase_id)
        await purchase_ref.set(purchase_data.__dict__)
        await self.update_user_with_purchase(purchase_ref)

    async def update_purchase(self, purchase_data: Dict[str, Any]):
        purchases = self.firestore.collection('purchases')
        purchase_id = f"{purchase_data['iapSource']}_{purchase_data['orderId']}"
        purchase_ref = purchases.document(purchase_id)
        await purchase_ref.update(purchase_data)
        await self.update_user_with_purchase(purchase_ref)

    async def update_user_with_purchase(self, purchase_ref):
        purchase_doc = await purchase_ref.get()
        purchase_data = purchase_doc.to_dict()
        if 'userId' in purchase_data:
            user_ref = self.firestore.collection('users').document(purchase_data['userId'])
            if purchase_data['productId'] in ['unlimited_yt_monthly', 'unlimited_yt_yearly']:
                if purchase_data['status'] == 'ACTIVE':
                    await user_ref.update({'unlimited': True})
                elif purchase_data['status'] == 'EXPIRED':
                    await user_ref.update({'unlimited': False})

    async def expire_subscriptions(self):
        documents = self.firestore.collection('purchases').where('expiryDate', '<=', firestore.SERVER_TIMESTAMP).where('status', '==', 'ACTIVE').stream()
        write_batch = self.firestore.batch()
        for doc in documents:
            data = doc.to_dict()
            if data['productId'] in ['unlimited_yt_monthly', 'unlimited_yt_yearly']:
                user_ref = self.firestore.collection('users').document(data['userId'])
                await user_ref.update({'unlimited': False})
            write_batch.update(doc.reference, {'status': 'EXPIRED'})
        await write_batch.commit()

# Example usage:
firebase_calls = FirebaseCalls(db)

File: payment_verification_backend/src/firebase.calls.ts
import { firestore } from "firebase-admin";
import Timestamp = firestore.Timestamp;

export type SubscriptionStatus = "PENDING" | "ACTIVE" | "EXPIRED";
export type NonSubscriptionStatus = "PENDING" | "COMPLETE" | "CANCELED";

export type IAPSource = "google_play" | "app_store";
export type Purchase = SubscriptionPurchase | NonSubscriptionPurchase;

export interface BasePurchase {
  iapSource: IAPSource;
  orderId: string;
  productId: string;
  userId: string;
  purchaseDate: firestore.Timestamp;
}

export interface SubscriptionPurchase extends BasePurchase {
  type: "SUBSCRIPTION";
  expiryDate: firestore.Timestamp;
  status: SubscriptionStatus;
}

export interface NonSubscriptionPurchase extends BasePurchase {
  type: "NON_SUBSCRIPTION";
  status: NonSubscriptionStatus;
}

export class FirebaseCalls {
  constructor(private firestore: FirebaseFirestore.Firestore) {}

  async createOrUpdatePurchase(purchaseData: Purchase): Promise<void> {
    const purchases = this.firestore.collection("purchases");
    const purchaseId = `${purchaseData.iapSource}_${purchaseData.orderId}`;
    const purchase = purchases.doc(purchaseId);
    await purchase.set(purchaseData);
    this.updateUserWithPurchase(purchase);
  }

  async updatePurchase(
    purchaseData: { iapSource: IAPSource; orderId: string } & Partial<Purchase>
  ): Promise<void> {
    const purchases = this.firestore.collection("purchases");
    const purchaseId = `${purchaseData.iapSource}_${purchaseData.orderId}`;
    const purchase = purchases.doc(purchaseId);
    await purchase.update(purchaseData);
    this.updateUserWithPurchase(purchase);
  }

  async updateUserWithPurchase(
    purchaseRef: firestore.DocumentReference
  ): Promise<void> {
    const purchaseDoc = await purchaseRef.get();
    if (typeof purchaseDoc?.data()?.userId != undefined) {
      const userRef = this.firestore
        .collection("users")
        .doc(purchaseDoc?.data()?.userId);
      if (!["1m", "1year"].includes(purchaseDoc?.data()?.productId)) return;
      if (purchaseDoc?.data()?.status == "ACTIVE") {
        userRef.update({
          premium: true,
        });
      } else if (purchaseDoc?.data()?.status == "EXPIRED") {
        userRef.update({
          premium: false,
        });
      }
    }
  }

  async expireSubscriptions(): Promise<void> {
    const documents = await this.firestore
      .collection("purchases")
      .where("expiryDate", "<=", Timestamp.now())
      .where("status", "==", "ACTIVE")
      .get();
    if (!documents.size) return;
    const writeBatch = this.firestore.batch();
    documents.docs.forEach((doc) => {
      if (["1m", "1year"].includes(doc.data().productId)) {
        const userRef = this.firestore
          .collection("users")
          .doc(doc.data().userId);
        userRef.update({
          premium: false,
        });
      }
      writeBatch.update(doc.ref, { status: "EXPIRED" });
    });
    await writeBatch.commit();
  }
}
